learning-another-language/variables.md

```diff
@@ -0,0 +1,200 @@
+| ---------------- | -------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------ |
+| Declaring        | Stating a variable's name and the type of variable it is.                                                                  | "We need to declare the variable before we assign a value to it," "We only need to declare a variable once." |
+| Assigning        | Creating a reference between a variable and an object.                                                                     | "I assigned the value `42` to the `let` variable named `jingsFavNum`."                                       |
+| `let` variable   | A variable that is block-scoped and re-assignable.                                                                         |
+| `const` variable | A constant variable. This variable is block-scoped and read-only. Constants are the type of variable we'll use most often. |
+
+## Declare Variables, Then Assign (Then Maybe Re-Assign)
+
+In JavaScript, there are two distinct steps when dealing with variables:
+
+1. declaration
+1. assignment
+
+**Declaration** is the step of stating a variable's name and the type of variable it is.
+
+Different variable types determine the variable's scope and re-assignment properties. Details are covered below!
+
+This is an example of variable declaration:
+
+```javascript
+let myFavoriteGreeting;
+```
+
+**Assignment** is the step of assigning a value to the variable, using the assignment operator `=`. Re-assignment also uses the assignment operator.
+
+```javascript
+let myFavoriteGreeting; // <- declaration
+myFavoriteGreeting = "Hello, World!"; // <- assignment
+myFavoriteGreeting = "Hello, Internet!"; // <- re-assignment
+```
+
+We only need to declare a variable once. We never need to re-declare it.
+
+### Declaration and Assignment Together
+
+We can do declaration and initial assignment on one line:
+
+```javascript
+let clicheGoodbye = "Don't forget to like and subscribe";
+```
+
+## Types of Variables
+
+All variables must be declared before we assign values to them.
+
+Variables are declared as one of three types:
+
+| Keyword | Definition                                                                                         |
+| ------- | -------------------------------------------------------------------------------------------------- |
+| `let`   | A block-scoped variable. Actively encouraged.                                                      |
+| `const` | A block-scoped variable that is read-only. Reassignment raises a `TypeError`. Actively encouraged. |
+| `var`   | A globally-scoped variable. Actively discouraged.                                                  |
```

Comment:
In the upcoming example, it's only global because we are in global scope. In a function, var implies function scope (not block scope) with hoisting semantics.

Reply:
I'll return to this-- It's worth correcting that it's not globally-scoped, and obviously you've seen a deliberate choice in not getting into hoisting

Reply:
Yeah. I don't think we'd want to get into the hoisting side. I didn't feel comfortable explicitly comparing to Python, though aside from the hoisting, it's a super apt comparison (and even Python has some surprising behavior in this regard!).

Did I really not make a suggestion here? Sorry for dropping the ball. Here's at least a start...

```suggestion
| `var`   | A variable without block scoping. If defined within a function, it will have scope throughout the _entire_ function. If defined outside a function it will have global scope. Actively discouraged.                                                  |
```